core/src/main/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryParser.java

```diff
@@ -120,160 +102,121 @@ public Query parse(QueryShardContext context) throws IOException, QueryParsingEx
             } else {
                 while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
                     if (token == XContentParser.Token.FIELD_NAME) {
                         currentFieldName = parser.currentName();
-                    } else if (parseContext.parseFieldMatcher().match(currentFieldName, WEIGHT_FIELD)) {
+                    } else if (parseContext.parseFieldMatcher().match(currentFieldName, FunctionScoreQueryParser.WEIGHT_FIELD)) {
                         functionWeight = parser.floatValue();
                     } else {
                         if ("filter".equals(currentFieldName)) {
-                            filter = parseContext.parseInnerFilter();
+                            filter = parseContext.parseInnerFilterToQueryBuilder();
                         } else {
-                            // do not need to check null here,
-                            // functionParserMapper throws exception if parser
-                            // non-existent
-                            ScoreFunctionParser functionParser = functionParserMapper.get(parseContext, currentFieldName);
-                            scoreFunction = functionParser.parse(context, parser);
+                            scoreFunction = copyFunctionSource(parser, currentFieldName);
                         }
                     }
                 }
+
                 if (functionWeight != null) {
-                    scoreFunction = new WeightFactorFunction(functionWeight, scoreFunction);
+                    if (scoreFunction == null) {
+                        //if only weight is present, that is the score function
+                        scoreFunction = new WeightBuilder().setWeight(functionWeight).buildAsBytes(XContentType.JSON);
+                    } else {
+                        //if both weight and function are present, we need to combine them
+                        final BytesReference function = scoreFunction;
+                        scoreFunction = new ScoreFunctionBuilder() {
+                            @Override
+                            protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+                                XContentParser parser = XContentFactory.xContent(function).createParser(function);
+                                parser.nextToken();
+                                //skip the start_object token from the function, we have to copy only starting from the name of the function
+                                parser.nextToken();
+                                builder.copyCurrentStructure(parser);
+                            }
+
+                            @Override
+                            public String getName() {
+                                throw new UnsupportedOperationException();
+                            }
+                        }.setWeight(functionWeight).buildAsBytes(XContentType.JSON);
+                    }
```

Comment:
sure, but sit tight, long story :)

`weight` can be its own function, note it's just a field at the same level as filter e.g.

```
{
    "functions" : [
        {
            "filter" : { "match_all" : {} },
            "weight" : 5
        }
    ]
}
```


or just a weight associated with another function e.g.

```
{
    "functions" : [
        {
            "filter" : { "match_all" : {} },
            "random_score" : {},
            "weight" : 5
        }
    ]
}
```

Internally, `weight` is treated as a special function (it is not registered to the ordinary score functions). What we have to do is try and copy the function, note that copyCurrentStructure will stop at the end of random_score. Then we have to see what we find. If we find another function we  throw error (something I am adding as we speak), otherwise we could find the weight. We parse the weight, but it needs to be associated with the existing function after the fact. Meaning that we have to reparse the existing function, add the weight to it and make the two functions a single funcion. I hope this explains things... it took me a while to figure this out...